File: bisection.py
import error
prevError=error.trueError(3,1)
def fun(x):
    return x*x*x - x*x + 5

def bisection(a, b):
    if fun(a) * fun(b) >= 0:
        print("Root doesn't exist in this interval.")
        return

    prev_c = 0
    iterator = 1

    print("I\t a\t b\t c\t error")
    
    while True:
        c = (a + b) / 2
        error = abs((c - prev_c) / c) * 100
        print(f"{iterator}\t {a:.2f}\t {b:.2f}\t {c:.2f}\t {error:.2f}")

        if fun(c) == 0.0:
            break
        elif fun(c) * fun(a) > 0:
            a = c
        else:
            b = c

        prev_c = c
        iterator += 1

        if error <prevError:
            break

    print("The value of root is:", c)


bisection(-5,3)

File: bisection.java
public class BisectionMethod {
    // Function f(x)
    public static double f(double x) {
        return 3 * x - Math.cos(x) - 1;
    }

    // Bisection Method
    public static void bisection(double a, double b, double tolerance) {
        if (f(a) * f(b) >= 0) {
            System.out.println("Root does not exist in the given interval.");
            return;
        }

        double c = 0, error = 1;
        while (error > tolerance) {
            c = (a + b) / 2;  // Midpoint
            error = Math.abs(f(c));  // Error as absolute value of f(c)

            if (f(c) == 0) {
                break;  // c is the root
            } else if (f(a) * f(c) < 0) {
                b = c;  // Root lies between a and c
            } else {
                a = c;  // Root lies between c and b
            }
        }

        System.out.println("The root is: " + c);
    }

    public static void main(String[] args) {
        bisection(0, 1, 0.001);  // Interval [0, 1] and tolerance 0.001
    }
}
